src/main/java/org/logstash/tcp/InputLoop.java

```diff
@@ -0,0 +1,198 @@
+package org.logstash.tcp;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.StandardSocketOptions;
+import java.nio.ByteBuffer;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.nio.channels.ServerSocketChannel;
+import java.nio.channels.SocketChannel;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+public final class InputLoop implements Runnable, Closeable {
+
+    /**
+     * Read buffer size.
+     */
+    private static final int BUFFER_SIZE = 8096;
+
+    /**
+     * Queue of {@link SelectionKey} that have data attached and are waiting to be drained by a
+     * JRuby thread.
+     */
+    private final ArrayBlockingQueue<SelectionKey> readable = new ArrayBlockingQueue<>(64);
```

Comment:
@jakelandis yes, but think about this:

Every key that you put in here becomes inactive for receiving data (this could be further improved by using buffer pairs, but I didn't want to go crazy just yet :)). We are only draining this queue using a single thread, 64 already seems like a lot actually.
Guess it depends on your codec and how burst heavy the incoming data is.
To me it seems like this shouldn't make much of a difference with internal socket buffers. We could make it larger though at no cost I think. `1024` would probably look the same and only consume a trivial amount of added memory :)?